HTTP: implement trailing headers for chuncked transfers

aneutron opened the pull request:
This patch adds the **CURLOPT_TRAILERDATA** and **CURLOPT_TRAILERFUNCTION** options that allow a callback based approach to sending trailing headers with chunked transfers. 

Following the _everything is a state machine_ mantra, it's four stages state machine that traps right before sending the final CRLF, calls the trailer function that fills a *curl_slist* with the trailers, compiles them into a single buffer, and proceeds to read from the buffer as if reading from the read callback, then jumps back into the next main state as if nothing happened.

The test server (sws) was updated to take into account the detection of the end of transfer in the case of trailing headers presence.

Test 1591 checks that trailing headers can be sent using libcurl.

_(I'm very sorry I had to remake this PR, [the first one](https://github.com/curl/curl/pull/2841) was made to merge my master branch into the curl master branch, it was a rookie mistake that made rebasing difficult ... That and the lack of tests, half-baked docs and tests that didn't pass. Suffice to say that mistakes were made. I sincerely apologize. All the remarks on the previous PR were addressed.)_

bagder commented on docs/libcurl/opts/CURLOPT_HTTP_TRAILER_DATA.3:
```diff
@@ -0,0 +1,58 @@
+.\" *                            | (__| |_| |  _ <| |___
+.\" *                             \___|\___/|_| \_\_____|
+.\" *
+.\" * Copyright (C) 1998 - 2017, Daniel Stenberg, <daniel@haxx.se>, et al.
```
@danielgustafsson it seems the travis check for copyright year mismatch doesn't work on new files?

bagder commented on docs/libcurl/opts/CURLOPT_HTTP_TRAILER_DATA.3:
```diff
@@ -0,0 +1,58 @@
+.TH CURLOPT_HTTP_TRAILER_DATA 3 "14 Aug 2018" "libcurl 7.??.?" "curl_easy_setopt options"
+
+.SH NAME:
+CURLOPT_HTTP_TRAILER_DATA \- Set user data to be passed on to the trailing headers callback 
```
"custom pointer passed to the trailing headers callback" maybe? We want it to be short and sweet. It also makes sense to have all *DATA options described using similar wording.

bagder commented on docs/libcurl/curl_easy_setopt.3:
```diff
@@ -327,6 +327,12 @@ Disable Content decoding. See \fICURLOPT_HTTP_CONTENT_DECODING(3)\fP
 Disable Transfer decoding. See \fICURLOPT_HTTP_TRANSFER_DECODING(3)\fP
 .IP CURLOPT_EXPECT_100_TIMEOUT_MS
 100-continue timeout. See \fICURLOPT_EXPECT_100_TIMEOUT_MS(3)\fP
+.IP CURLOPT_HTTP_TRAILER_FUNCTION
```
A little bikeshedding here perhaps, but...

I'm not convinced it is a good idea to include `HTTP` in the name. What if there's another protocol with a similar concept that we could use this callback for at a later point in time? Maybe just `CURLOPT_TRAILERFUNCTION`  ? Most of the existing *FUNCTION names are already named without a separating underscore.

bagder commented on docs/libcurl/opts/CURLOPT_HTTP_TRAILER_DATA.3:
```diff
@@ -0,0 +1,58 @@
+
+.SH EXAMPLE:
+.nf
+// Assuming we have a CURL handle in the hndl variable.
```
Since we discourage `//` comments everywhere (for C89 portability), we probably should use them in the man page example either...

bagder commented on docs/libcurl/opts/CURLOPT_HTTP_TRAILER_FUNCTION.3:
```diff
@@ -0,0 +1,110 @@
+.TH CURLOPT_HTTP_TRAILER_FUNCTION 3 "14 Aug 2018" "libcurl 7.63.0" "curl_easy_setopt options"
+
+.SH NAME:
+CURLOPT_HTTP_TRAILER_FUNCTION \- Callback to fill trailing headers list to be sent in a chunked transfer.
```
Here too I think this should be as short as possible that still conveys what it does:

"set callback for sending trailing headers" perhaps?

bagder commented on docs/libcurl/opts/CURLOPT_HTTP_TRAILER_FUNCTION.3:
```diff
@@ -0,0 +1,110 @@
+.SH SYNOPSIS:
+#include <curl.h>
+
+void curl_trailer_callback(struct curl_slist ** list, void *userdata);
```
Please make this callback able to return an error if it runs into problems. I would recommend using return 0 for no problems, and perhaps 1 to abort the request and keep all other values reserved for the future.

bagder commented on docs/libcurl/opts/CURLOPT_HTTP_TRAILER_FUNCTION.3:
```diff
@@ -0,0 +1,110 @@
+
+static void trailer_cb(struct curl_slist **tr, void *data)
+{
+    /* libcurl will free the list as soon as it becomes not needed */
```
"libcurl will free the list" will be enough and avoids complicating the statement

bagder commented on docs/libcurl/opts/CURLOPT_HTTP_TRAILER_FUNCTION.3:
```diff
@@ -0,0 +1,110 @@
+    curl_easy_setopt(curl, CURLOPT_READFUNCTION, read_cb);
+
+    /* We need to announce the trailers to come 
+       See CURLOPT_READFUNCTION for more on the matter */
```
But this isn't explained there, is it?

aneutron replied:
This is a weird combined leftover from a previous version of the patch and a copy pasted example that was modified. It's not required at all. I'm really sorry for this inconsistency.

bagder commented on docs/libcurl/opts/CURLOPT_HTTP_TRAILER_DATA.3:
```diff
@@ -0,0 +1,58 @@
+
+A more complete example can be found in examples/http_trailers.html
+.SH AVAILABILITY:
+This option was added in curl 7.63.0 and is present if HTTP support is enabled
```
It won't have time to get into 7.63.0, but presumable 7.64.0...

bagder commented on docs/libcurl/symbols-in-versions:
```diff
@@ -431,6 +431,8 @@ CURLOPT_HTTPREQUEST             7.1           -           7.15.5
 CURLOPT_HTTP_TRANSFER_DECODING  7.16.2
 CURLOPT_HTTP_VERSION            7.9.1
+CURLOPT_HTTP_TRAILER_FUNCTION   7.63.2
+CURLOPT_HTTP_TRAILER_DATA       7.63.2
```
Most likely this can go into 7.64.0

bagder commented on lib/transfer.c:
```diff
@@ -140,15 +179,63 @@ CURLcode Curl_fillreadbuffer(struct connectdata *conn, size_t bytes,
+  /* if we are transmitting trailing data, we don't need to write
+     a chunk size so we skip this */
+  if(data->req.upload_chunky
+     && data->state.trailers_state == HTTP_TRAILERS_NONE) {
```
When splitting long expressions, we usually prefer to split the lines with the operator on the previous line, and I would recommend parentheses:

    (data->req.upload_chunky &&
     (data->state.trailers_state == HTTP_TRAILERS_NONE))

bagder commented on lib/transfer.c:
```diff
@@ -251,13 +354,29 @@ CURLcode Curl_fillreadbuffer(struct connectdata *conn, size_t bytes,
+    else
+#endif
+      if((nread - hexlen) == 0 &&
+            data->state.trailers_state != HTTP_TRAILERS_INITIALIZED) {
```
this looks funnily indented

bagder commented on lib/urldata.h:
```diff
@@ -1216,6 +1216,15 @@ typedef enum {
 
+
+typedef enum {
+    HTTP_TRAILERS_NONE,
```
not two-space indentation

bagder commented on lib/urldata.h:
```diff
@@ -1362,6 +1371,13 @@ struct UrlState {
   CURLU *uh; /* URL handle for the current parsed URL */
   struct urlpieces up;
+#ifndef CURL_DISABLE_HTTP
+  size_t trailers_bytes_sent; /* no other way of keeping track of sent bytes */
```
no point in explaining what it *isn't*...

bagder commented on lib/transfer.c:
```diff
@@ -117,6 +117,35 @@ CURLcode Curl_get_upload_buffer(struct Curl_easy *data)
+{
+  struct Curl_easy *data = (struct Curl_easy *)raw;
+  Curl_send_buffer *trailers_buf = data->state.trailers_buf;
+  return trailers_buf->size_used-data->state.trailers_bytes_sent;
```
maybe add spaces to make this more readable?

    return trailers_buf->size_used - data->state.trailers_bytes_sent;

aneutron commented:
I think I addressed all the remarks:
- Replaced CULROPT_HTTP_TRAILER* with CURLOPT_TRAILER*
- Renamed the trailer state enum members from HTTP_TRAILER_* to TRAILER_*
- Added two defs _(CURL_TRAILERFUNC_OK & CURL_TRAILER_FUNC_ABORT)_ to be used as return codes for the trailer function that now returns an integer.
- Fixed the dates and releases on the docs, and edited the docs as suggested.
- Fixed the weird indentations that slipped.
- _(I'm sorry I just noticed this a day after I addressed everything else)_ Fixed the commit message to reflect these changes.

mkauf commented:
Please fix the typo in the commit message: "chuncked" --> "chunked".

aneutron commented:
@mkauf Should I leave `HTTP:` in the commit message ?

bagder commented:
> Should I leave HTTP: in the commit message?

Yes I think so, since this is after all only for HTTP (now).

bagder commented:
Thanks!
